### DISCUSSION TITLE
Is it possible to implement slope limiters in DG kernels?

### DISCUSSION THREAD

**Discussion by ngrilli** [idaholab/moose#27355] [Q&A General]
Hello MOOSE Users,

I have previously implemented dislocation transport equations using DGKernels for advection. 
These kernels were an extension of DGConvection with custom velocity directions 
that depend on the crystallographic directions of my problem.

An example can be found here:
https://github.com/ngrilli/c_pfor_am/blob/main/src/dgkernels/DGAdvectionCoupled.C

Now, I would like to add a slope limiter to mitigate some of the oscillations 
at sharp interfaces of the dislocation density variable. 

I have read the finite volume code available, 
such as FVAdvection and FVLimitedAdvection, 
where slope limiters can be easily introduced.

I am asking whether it is possible to incorporate slope limiters into DGConvection 
and if anyone has explored this. I could not find any examples of this approach
in the framework or modules.

I am at an early stage of this development and would appreciate your suggestion: 
should I implement my previous transport laws into the FV framework, 
or is there a simple way to retrofit my previous DG kernels with slope limiters?

I don't mind about the order of the variables because I previously used the DGKernels
with constant monomial variables, therefore it was behaving like FV anyway.
I am just looking for the easiest way to implement it.

Thank you very much in advance for your help.

Best Regards,
Nicolò

**Comment by GiudGiud** [comment]
Hello

This has been done in the thermal hydraulics module for 1D flow. 
look at SlopeReconstruction1DInterface and the RDG3EqnMaterial.
DGKernels are also used for this variable-area Euler equation implementation. 

I dont think anyone has done this for DGConvection. @lindsayad if you've heard of this. 

> should I implement my previous transport laws into the FV framework,
> I don't mind about the order of the variables because I previously used the DGKernels
> with constant monomial variables, therefore it was behaving like FV anyway.

I would wager writing code in FV is easier than DG. We are pouring a lot of resources into FV. We have better success there than we have ever had with CG. DG has not been explored fully imo, and @lindsayad is exploring Hybridized DG in MOOSE currently

Guillaume

**Comment by ngrilli** [comment]
Thank you very much @GiudGiud 

I will try the implementation with FV,
after a quick read of the code, the FV looks easier for me as well.

Nicolò

**Reply by GiudGiud** [reply]
sounds good. FV already has slope limiters implemented in a utils, used in the MathFVUtils

**Comment by lindsayad** [comment]
> I am at an early stage of this development and would appreciate your suggestion:
should I implement my previous transport laws into the FV framework,
or is there a simple way to retrofit my previous DG kernels with slope limiters?

If you're constraining yourself to at most a second order accurate (L2 norm) method, then I agree that FV is the easiest since it is *the* second order accurate reconstructed DG method. Did you have a reference for a particular slope limiter? It seems like maybe not
